src/bundle/bundleRouter.ts

```diff
@@ -13,46 +17,119 @@ export interface IBundleRouteProps {
+    let fileName = beautifyBundleName(paths.shift());
+    while (splitFileNames[fileName]) {
+      fileName = beautifyBundleName(paths.shift() + path.sep + fileName);
+    }
+    splitFileNames.push(fileName);
+    return fileName;
   }
 
-  function dispatch(module: IModule) {
-    const bundle = getBundle(module);
+  function createMainBundle() {
+    mainBundle = createBundle({
+      bundleConfig: outputConfig.app,
+      ctx: ctx,
+      entries,
+      includeAPI: true,
+      priority: 2,
+      type: BundleType.JS_APP,
+    });
+  }
 
+  function createVendorBundle() {
+    vendorBundle = createBundle({
+      bundleConfig: outputConfig.vendor,
+      ctx: ctx,
+      includeAPI: false,
+      priority: 1,
+      type: BundleType.JS_VENDOR,
+    });
+  }
+
+  function dispatch(bundle: Bundle, module: IModule) {
     if (!module.isCached) {
       ict.sync('bundle_resolve_module', { module: module });
     }
     bundle.source.modules.push(module);
   }
 
-  const scope: IBundleRouter = {
-    dispatchModules: (modules: Array<IModule>, productionContext?: IProductionContext) => {
-      // console.log(productionContext.splitEntries);
+  const self: IBundleRouter = {
+    generateBundles: (modules: Array<IModule>) => {
       for (const module of modules) {
-        // console.log(module.id);
-        dispatch(module);
+        // we skip this module
+        if (module.isSplit) {
+          continue;
+        } else if (module.pkg.type === PackageType.EXTERNAL_PACKAGE && hasVendorConfig) {
+          if (!vendorBundle) createVendorBundle();
+          dispatch(vendorBundle, module);
+        } else {
+          if (!mainBundle) createMainBundle();
+          dispatch(mainBundle, module);
+        }
+      }
+    },
+    generateSplitBundles: (entries: Array<ISplitEntry>) => {
+      for (const splitEntry of entries) {
+        const { entry, modules } = splitEntry;
+        const fileName = generateSplitFileName(entry.publicPath);
+        // @todo: improve the codeSplitting config
+        const splitBundle = createBundle({
+          bundleConfig: {
+            path: outputConfig.codeSplitting.path,
+          },
+          ctx,
+          fileName,
+          includeAPI: false,
+          type: BundleType.JS_SPLIT,
+          webIndexed: false,
+        });
+        for (const module of modules) {
+          dispatch(splitBundle, module);
+        }
+        const bundleConfig = splitBundle.prepare();
+        codeSplittingMap.entries[entry.id] = {
+          b: bundleConfig.relativePath,
+        };
+        bundles.push(splitBundle);
       }
     },
     writeBundles: async () => {
-      return await mainBundle.generate();
+      let output = [];
+      if (Object.keys(codeSplittingMap.entries).length > 0) {
+        mainBundle.addCodeSplittingMap(codeSplittingMap);
+      }
+      await Promise.all([mainBundle.generate(), vendorBundle.generate(), ...bundles.map(bundle => bundle.generate())])
```

Comment:
During development we most like will have only one bundle. I am afraid that this will be a performance bottleneck + vendorBundle could be undefined